plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,278 @@
+   work with closures instead of terms; this also makes it easier
+   to ensure that we're using the correct environment. -}
+
+
+-- | Stack frames.  These are effectively continuations: would using explicit CPS speed things up?
+-- The spec also has a frame for n-ary built in application, but that's not in the AST yet
+data Frame
+    = FrameAppArg Closure                        -- ^ @[_ N]@       -- Tells the machine that we've got back to N after evaluating M in [M N]
+    | FrameHeapUpdate HeapLoc                    -- ^ @(update l)@  -- Mark the stack for evaluation of a delayed argument
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+      deriving (Show)
+
+-- | A context is a stack of frames.
+type EvaluationContext = [Frame]
+
+-- | A local version of the EvaluationResult type using closures instead of values
+data LMachineResult
+    = Success Closure Heap  -- We need both the environment and the heap to see what the value "really" is.
+    | Failure
+
+emptyEnvironment :: Environment
+emptyEnvironment = Environment IntMap.empty
+
+-- | Extend an environment with a new binding
+updateEnvironment :: Int -> HeapLoc -> Environment -> Environment
+updateEnvironment index cl (Environment m) = Environment (IntMap.insert index cl m)
+
+-- | Look up a heap location in an environment.
+lookupHeapLoc :: Name () -> Environment -> HeapLoc
+lookupHeapLoc name (Environment env) =
+    case IntMap.lookup (unUnique $ nameUnique name) env of
+      Nothing  -> throw $ LMachineStringException ("Name " ++ show (nameString name) ++ " missing from environment")
+      Just loc -> loc
+
+emptyHeap :: Heap
+emptyHeap = Heap IntMap.empty 0
+
+-- | Insert a new entry into a heap. The heap grows monotonically with no garbage collection or reuse of heap slots.
+insertInHeap :: Closure -> Heap -> (HeapLoc, Heap)
+insertInHeap cl (Heap h top) =
+    let top' = top + 1
+    in (top', Heap (IntMap.insert top' (Unevaluated cl) h) top')
+
+-- | Replace the heap entry at a given location.
+-- The old entry presumably persits in the old heap.
+updateHeap :: HeapLoc -> HeapEntry -> Heap -> Heap
+updateHeap l cl (Heap h top) =
+    Heap (IntMap.insert l cl h) top
+    -- insert seems to be faster than adjust (18s vs 28 s for fib 32) and also uses less memory.
+
+lookupHeap :: HeapLoc -> Heap -> HeapEntry
+lookupHeap l (Heap h _) =
+    case IntMap.lookup l h of
+      Just e  -> e
+      Nothing -> throw $ LMachineStringException ("Missing heap location in lookupHeap: " ++ show l)  -- This should never happen
+
+
+-- | The basic computation step of the L machine.  Search down the AST looking for a value, saving surrounding contexts on the stack.
+computeL :: EvaluationContext -> Heap -> Closure -> LMachineResult
+computeL ctx heap cl@(Closure term env) =
+    case term of
```

Comment:
I'd probably put this on the previous line, saving you an indent on the rest of the function.

Reply:
Me too.